tfhe/src/core_crypto/gpu/slice.rs

```diff
@@ -0,0 +1,194 @@
+use crate::core_crypto::gpu::vec::range_bounds_to_start_end;
+use crate::core_crypto::gpu::{CudaPtr, CudaStream};
+use crate::core_crypto::prelude::Numeric;
+use std::ffi::c_void;
+use std::marker::PhantomData;
+use tfhe_cuda_backend::cuda_bind::{cuda_memcpy_async_gpu_to_gpu, cuda_memcpy_async_to_cpu};
+
+#[derive(Debug)]
+pub struct CudaSlice<'a, T: Numeric> {
+    ptr: CudaPtr,
+    _len: usize,
+    _phantom_1: PhantomData<T>,
+    _phantom_2: PhantomData<&'a ()>,
+}
+
+#[derive(Debug)]
+pub struct CudaSliceMut<'a, T: Numeric> {
+    ptr: CudaPtr,
+    len: usize,
+    _phantom_1: PhantomData<T>,
+    _phantom_2: PhantomData<&'a mut ()>,
+}
+
+impl<'a, T> CudaSlice<'a, T>
+where
+    T: Numeric,
+{
+    /// # Safety
+    pub(crate) unsafe fn new(ptr: CudaPtr, len: usize) -> CudaSlice<'a, T> {
+        CudaSlice {
+            ptr,
+            _len: len,
+            _phantom_1: PhantomData,
+            _phantom_2: PhantomData,
+        }
+    }
+
+    /// # Safety
+    pub(crate) unsafe fn as_c_ptr(&self) -> *const c_void {
+        self.ptr.as_c_ptr()
+    }
+}
+
+impl<'a, T> CudaSliceMut<'a, T>
+where
+    T: Numeric,
+{
+    pub(crate) unsafe fn new(ptr: CudaPtr, len: usize) -> CudaSliceMut<'a, T> {
+        CudaSliceMut {
+            ptr,
+            len,
+            _phantom_1: PhantomData,
+            _phantom_2: PhantomData,
+        }
+    }
+
+    /// # Safety
```

Comment:
We can add 
```
The caller must ensure that the slice outlives the pointer this function returns, or else it will end up pointing to garbage.
```
(Stolen from Rust's stdlib)